pyomo/repn/parameterized_linear.py

```diff
@@ -0,0 +1,400 @@
+_GENERAL = ParameterizedExprType.GENERAL
+
+
+def _merge_dict(dest_dict, mult, src_dict):
+    if mult.__class__ not in native_numeric_types or mult != 1:
+        for vid, coef in src_dict.items():
+            if vid in dest_dict:
+                dest_dict[vid] += mult * coef
+            else:
+                dest_dict[vid] = mult * coef
+    else:
+        for vid, coef in src_dict.items():
+            if vid in dest_dict:
+                dest_dict[vid] += coef
+            else:
+                dest_dict[vid] = coef
+
+
+def to_expression(visitor, arg):
+    if arg[0] in (_CONSTANT, _PSEUDO_CONSTANT):
+        return arg[1]
+    else:
+        return arg[1].to_expression(visitor)
+
+
+class ParameterizedLinearRepn(LinearRepn):
+    def __str__(self):
+        return (
+            f"ParameterizedLinearRepn(mult={self.multiplier}, const={self.constant}, "
+            f"linear={self.linear}, nonlinear={self.nonlinear})"
+        )
+
+    def walker_exitNode(self):
+        if self.nonlinear is not None:
+            return _GENERAL, self
+        elif self.linear:
+            return _LINEAR, self
+        elif self.constant.__class__ in native_numeric_types:
+            return _CONSTANT, self.multiplier * self.constant
+        else:
+            return _PSEUDO_CONSTANT, self.multiplier * self.constant
+
+    def to_expression(self, visitor):
+        if self.nonlinear is not None:
+            # We want to start with the nonlinear term (and use
+            # assignment) in case the term is a non-numeric node (like a
+            # relational expression)
+            ans = self.nonlinear
+        else:
+            ans = 0
+        if self.linear:
+            var_map = visitor.var_map
+            with mutable_expression() as e:
+                for vid, coef in self.linear.items():
+                    if coef.__class__ not in native_numeric_types or coef:
+                        e += coef * var_map[vid]
+            if e.nargs() > 1:
+                ans += e
+            elif e.nargs() == 1:
+                ans += e.arg(0)
+        if self.constant.__class__ not in native_numeric_types or self.constant:
+            ans += self.constant
+        if (
+            self.multiplier.__class__ not in native_numeric_types
+            or self.multiplier != 1
+        ):
+            ans *= self.multiplier
+        return ans
+
+    def append(self, other):
+        """Append a child result from acceptChildResult
+
+        Notes
+        -----
+        This method assumes that the operator was "+". It is implemented
+        so that we can directly use a LinearRepn() as a `data` object in
+        the expression walker (thereby allowing us to use the default
+        implementation of acceptChildResult [which calls
+        `data.append()`] and avoid the function call for a custom
+        callback).
+
+        """
+        _type, other = other
+        if _type in (_CONSTANT, _PSEUDO_CONSTANT):
```

Comment:
This is slower than necessary (it creates a tuple, does a linear scan, and throws it away each time).  Maybe
```
if _type is _CONSTANT or _type is _PSEUDO_CONSTANT:
```